mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java

```diff
@@ -687,13 +687,29 @@ private boolean hasChildIn(Mailbox parentMailbox, List<Mailbox> mailboxesWithPat
     }
 
     private Stream<MailboxId> getInMailboxes(ImmutableSet<MailboxId> inMailboxes, MailboxSession session) throws MailboxException {
-       if (inMailboxes.isEmpty()) {
+        if (inMailboxes.isEmpty()) {
             return getAllReadableMailbox(session);
         } else {
+            if (inMailboxes.size() == 1) {
+                MailboxId mailboxId = inMailboxes.stream().findAny().get();
+                return filterReadable(mailboxId, session);
+            }
             return getAllReadableMailbox(session).filter(inMailboxes::contains);
```

Comment:
Why don't we push the predicate down to allow the search to use it?
The optimization you did assumes things that only the implementation should know (like latency on `hasRight` vs latency on `searchMailboxes` for example)

Reply:
The Mapper api expects a condition on the name, not on the ids.

I'm unsure of the result you expect.

Reply:
This is a typical case when talking about search: you have a predicate (or a clause) for search and the backend don't know how to implement it.
Either you implement it un-efficiently at the top level and backends won't be able to optimize for that case: quick to implement but will need more work in the future.
Or you implement a method for that in the backends that actually do the same dumb filter but you open the door to an efficient implementation in backends.

I suggest we do the second one.

To talk about code, as you know, I'm not a big fan of `Mapper` concept.

We already have this nice search interface we need to push down the predicates, it's the `search` method.

Let's just create 3 protected methods:
* listAllMailboxes
* searchSingleMailbox
* searchMailboxesById

It allows any type extending StoreMailboxManager to optimize  some cases.